### PR TITLE
Plenarex merge

### PR DESCRIPTION
**Pull request by vforgione**
Before things diverge any further, this is the complete merge in of what I was working on with Plenarex and all of its trickle down changes. Obviously, it's a lot.

**Highlights:**

- the migrations have been consolidated and renamed to keep things nice and clean
- the schemas have been moved into their proper contexts
- actions are now more flexible with their params (guards and matching all over the place)
- the controllers have been wiped and redone using more phoenix-y idioms
- metas can now be searched against (roughly)

There's a lot going on in here, but overall it does the exact same thing as the existing code. It's just a shit load cleaner, standardized and more tolerant. Also, the web front end is prettier. I've also found and fixed a number of bugs and added in some missing functionality.

### CODE REVIEW

File: lib/plenario/actions/meta_actions.ex [hunk 1]
```diff
@@ -1,424 +1,388 @@
...
-    msg = "Data set error fixed:\n\n" <> message
-
-    AdminUserNoteActions.create_for_meta(
-      msg,
-      admin,
-      meta.user,
-      meta,
-      true
-    )
+  defp parse_json(filename, count) do
```
Comment (HeyZoos) [inline lib/plenario/actions/meta_actions.ex]: These parsing functions are really similar to those used in the worker. I'll make a note to abstract these out to their own modules when I get the chance.

File: lib/plenario/changesets/changeset_utils.ex [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  @name_length 16
+
+  def validate_meta_state(%Ecto.Changeset{valid?: true} = changeset) do
+    meta = MetaActions.get(get_field(changeset, :meta_id))
+    case meta.state do
+      "new" -> changeset
+      _ -> add_error(changeset, :base, "Data set is locked")
+    end
+  end
+  def validate_meta_state(changeset), do: changeset
```
Comment (HeyZoos) [inline lib/plenario/changesets/changeset_utils.ex]: Ah okay, so it's flat across the board. Can't change meta or anything related to it if it's not "new".

File: lib/plenario/changesets/meta_changesets.ex [hunk 3]
```diff
@@ -1,212 +1,182 @@
...
-    rr = get_field(changeset, :refresh_rate)
-
-    if Enum.member?(Meta.get_refresh_rate_values(), rr) do
-      changeset
-    else
-      changeset |> add_error(:refresh_rate, "Invalid value `#{rr}`")
+  defp validate_state(changeset), do: changeset
+
+  defp test_source_url(%Ecto.Changeset{valid?: true, changes: %{source_url: source_url}} = changeset) do
+    case HTTPoison.options(source_url) do
```
Comment (HeyZoos) [inline lib/plenario/changesets/meta_changesets.ex]: I don't think this is a good idea to test if a page exists. From what I've read not many websites bother to respond properly to an OPTIONS request. You might get a lot of false negatives. Is it not possible to inspect the status code of a GET or POST without waiting for the response body?

File: lib/plenario/actions/data_set_actions.ex [hunk 4]
```diff
@@ -1,285 +1,195 @@
...
-  defp create_trigger!(trigger_name, table_name, function_name) do
-    sql =
-      EEx.eval_file(
-        "#{@template_dir}/create_trigger.sql.eex",
-        [trigger_name: trigger_name, table_name: table_name, function_name: function_name],
-        trim: true
-      )
-
+  defp execute_sql!(sql) do
+    Logger.info(sql)
```
[the remark this answers is not in the document]
Comment (vforgione) [reply lib/plenario/actions/data_set_actions.ex]: I just wanted it in there to debug failures. Now it's:

```elixir
  defp execute_sql!(sql) do
    case Ecto.Adapters.SQL.query(Repo, sql) do
      {:ok, _} ->
        :ok

      {:error, error} ->
        Logger.error(error.postgres.message)
        Logger.error(sql)
        raise error
    end
  end
```

File: lib/plenario/actions/data_set_field_actions.ex [hunk 5]
```diff
@@ -1,85 +1,95 @@
...
-  Updates a DataSetField
-  """
-  @spec update(field :: DataSetField, params :: map) :: DataSetField
-  def update(field, params) do
-    Logger.info("Updating DataSetField: #{inspect(field)}, #{inspect(params)}")
+  Updates a given DataSetField's attributes.
 
-    DataSetFieldChangesets.update(field, params)
+  If the related Meta instance's state field is not "new" though, this
+  will wrror out -- you cannot add a new DataSetField to and active Meta.
```
[the remark this answers is not in the document]
Comment (vforgione) [reply lib/plenario/actions/data_set_field_actions.ex]: Yes. And I'll fix the wrror typos...

File: lib/plenario/actions/meta_actions.ex [hunk 6]
```diff
@@ -1,424 +1,388 @@
...
-  end
+  @doc """
+  Create a new instance of Meta in the database.
 
-  defp get_by_slug(slug, opts) do
-    Q.from_slug(slug)
-    |> Q.handle_opts(opts)
-    |> Repo.one()
+  If the related Meta instance's state field is not "new" though, this
+  will wrror out -- you cannot add a new Meta to and active Meta.
```
[the remark this answers is not in the document]
Comment (vforgione) [reply lib/plenario/actions/meta_actions.ex]: I templated the docs for the actions - this is just a simple `sed` thing. If you look at stuff like `DataSetField` and `VirtualDateField` you'll see the pattern.

File: lib/plenario/changesets/changeset_utils.ex (hunk 2 above)
Comment (vforgione) [reply lib/plenario/changesets/changeset_utils.ex]: Correct.

File: lib/plenario/actions/meta_actions.ex [hunk 7]
```diff
@@ -1,424 +1,388 @@
...
+    guesses =
+      for row <- first_thousand do
+        for {key, value} <- row do
+          cond do
+            check_boolean(value) -> {key, "boolean"}
+            check_integer(value) -> {key, "integer"}
+            check_float(value) -> {key, "float"}
+            check_date(value) -> {key, "timestamptz"}
+            check_json(value) -> {key, "jsonb"}
+            true -> {key, "text"}
```
[the remark this answers is not in the document]
Comment (vforgione) [reply lib/plenario/actions/meta_actions.ex]: Later. I'll make an issue for it.

File: lib/plenario/actions/meta_actions.ex (hunk 1 above)
Comment (vforgione) [reply lib/plenario/actions/meta_actions.ex]: Needs an issue.

**Pull request merged**